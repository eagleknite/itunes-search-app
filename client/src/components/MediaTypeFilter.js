// MediaTypeFilter component definition
function MediaTypeFilter({ selectedMedia, onMediaChange }) {

    // A predefined list of media types with their labels and values
    const mediaTypes = [
        { label: 'All', value: 'all' },
        { label: 'Movie', value: 'movie' },
        { label: 'Podcast', value: 'podcast' },
        { label: 'Music', value: 'music' },
        { label: 'Audiobook', value: 'audiobook' },
        { label: 'Short Film', value: 'shortFilm' },
        { label: 'TV Show', value: 'tvShow' },
        { label: 'Software', value: 'software' },
        { label: 'Ebook', value: 'ebook' }
    ];

    // Component's JSX rendering
    return (
        // Dropdown selector for media types
        <select
            // The currently selected media type
            value={selectedMedia}
            // When the selected media changes, call the passed onMediaChange function
            onChange={(e) => onMediaChange(e.target.value)}
            className="bg-gray-200 p-2 rounded-md"
        >
            {/* Loop through each media type and render it as an option in the dropdown */}
            {mediaTypes.map(type => (
                <option key={type.value} value={type.value}>
                    {type.label}
                </option>
            ))}
        </select>
    )
}

// Exporting the MediaTypeFilter component
export default MediaTypeFilter;
