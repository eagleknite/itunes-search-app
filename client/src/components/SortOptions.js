// Functional component to render sort options for search results
function SortOptions({ selectedSort, onSortChange }) {
  return (
    <div className="ml-4">
       
        <label htmlFor="sortOptions" className="mr-2">Sort By:</label>
        
        <select
            id="sortOptions"
            value={selectedSort}  // Current selected sort value
            onChange={(e) => onSortChange(e.target.value)}  // Handle sort option change
            className="border p-2 rounded-md"  // Styling for the dropdown
        >
            <option value="releaseDate">Release Date</option>
            <option value="trackPrice">Price</option>
            <option value="primaryGenreName">Genre</option>
            <option value="artistName">Artist</option>
            <option value="collectionName">Album</option>
        </select>
    </div>
  )
}

export default SortOptions;
