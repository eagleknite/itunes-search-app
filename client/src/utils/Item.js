function Item({ item, isFavorite, addToFavorites, removeFromFavorites, onItemClick }) {
    return (
        <div 
            // Styling for the item container, with hover effects and a rounded border
            className="border rounded-md p-3 relative hover:shadow-lg transition-shadow duration-300 cursor-pointer"
            onClick={onItemClick}
        >
            <img src={item.artworkUrl100} alt={item.trackName || item.collectionName} className="mx-auto mb-2 w-full" />
            
            <h3 className="text-center text-lg font-bold mb-2">{item.trackName || item.collectionName}</h3>
            
            <p className="text-center text-sm mb-2">{item.shortDescription}</p>
            {(item.collectionPrice || item.trackRentalPrice) && (
                <div className="flex justify-center items-center space-x-2 mb-2">
                    {item.collectionPrice && <span className="text-sm bg-blue-200 text-blue-700 p-1 rounded">Purchase: ${item.collectionPrice}</span>}
                    {item.trackRentalPrice && <span className="text-sm bg-yellow-200 text-yellow-700 p-1 rounded">Rent: ${item.trackRentalPrice}</span>}
                </div>
            )}

            {item.isStreamable && <span className="bg-green-500 text-white rounded-full p-1 text-xs absolute top-2 left-2">Streamable</span>}
            
            {(item.trackExplicitness === "explicit" || item.collectionExplicitness === "explicit") && 
                <span className="bg-red-500 text-white rounded-full p-1 text-xs absolute bottom-2 left-2">Explicit</span>
            }

            {isFavorite ?
                <button 
                    className="absolute top-2 right-2 bg-red-500 text-white rounded-full p-1 hover:bg-red-600"
                    onClick={(e) => { e.stopPropagation(); removeFromFavorites(item); }}  // Use stopPropagation to prevent triggering container's click event
                >
                    Remove
                </button> :
                <button 
                    className="absolute top-2 right-2 bg-blue-500 text-white rounded-full p-1 hover:bg-blue-600"
                    onClick={(e) => { e.stopPropagation(); addToFavorites(item); }}  // Use stopPropagation to prevent triggering container's click event
                >
                    Add
                </button>
            }
        </div>
    );
}

export default Item;
