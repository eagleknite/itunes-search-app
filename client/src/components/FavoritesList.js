// Necessary imports from Redux and local files
import { useDispatch, useSelector } from "react-redux";
import { removeFavorite } from "../store/favoritesSlice";
import Item from "../utils/Item";

// FavoritesList component definition
function FavoritesList() {
    // Get the list of favorite items from Redux state
    const favorites = useSelector(state => state.favorites);
    
    // Log the favorites for debugging purposes
    console.log("Favorites:", favorites);
    
    // Initialize dispatch function for Redux actions
    const dispatch = useDispatch();

    // Handler to remove an item from the favorites list
    const handleRemoveFromFavorites = (item) => {
        console.log("Removing:", item);
        dispatch(removeFavorite(item));
    };

    // Handler to open the item's link in a new tab when clicked
    const handleItemClick = (item) => {
        if (item.collectionViewUrl) {
            window.open(item.collectionViewUrl, "_blank");
        }
    };
    
    // JSX rendering of the FavoritesList component
    return (
        <div className="p-4">
            <h2 className="mb-4 text-2xl font-bold text-center">Favorites</h2>
            { /* Conditional rendering: Show a message if favorites list is empty */ }
            {favorites.length === 0 &&
                <p className="text-center mb-4 text-gray-600">
                    Your favorites list is empty. To add items, click on the "Add" button on each item in the search results.
                </p>
            }
            { /* Grid layout for displaying favorite items */ }
            <div className="grid grid-cols-1 md:grid-cols-2 lg:grid-cols-3 gap-4">
                {favorites.map(item => (
                    <Item 
                        key={item.trackId || item.collectionId} 
                        item={item}
                        isFavorite={true}
                        removeFromFavorites={handleRemoveFromFavorites}
                        onItemClick={() => handleItemClick(item)}
                    />
                ))}
            </div>
        </div>
    );
}

// Export the FavoritesList component
export default FavoritesList;

