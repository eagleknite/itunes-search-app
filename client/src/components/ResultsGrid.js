// Required dependencies and actions from slices
import { useDispatch, useSelector } from "react-redux";
import { setSelectedItem } from "../store/selectedItemSlice";
import { addFavorite, removeFavorite } from "../store/favoritesSlice";
import Item from "../utils/Item";

// ResultsGrid component definition
function ResultsGrid() {
    // Fetching search results and favorites from the store
    const results = useSelector(state => state.searchResults.items);
    const favorites = useSelector(state => state.favorites);
    
    // Redux dispatch function
    const dispatch = useDispatch();

    // Function to handle item click
    const handleItemClick = (item) => {
        // Set the selected item in the store
        dispatch(setSelectedItem(item));

        // If the item has a collectionViewUrl, open it in a new tab
        if (item.collectionViewUrl) {
            window.open(item.collectionViewUrl, "_blank");
        }
    };

    // Loop to check if all items have a trackId or collectionId (for audiobooks)
    results.forEach(item => {
        if (!item.trackId && !item.collectionId) {
            console.error("Item without trackId or collectionId:", item);
        }
    });

    // Function to handle adding item to favorites
    const handleAddToFavorites = (item) => {
        dispatch(addFavorite(item));
    };

    // Function to handle removing item from favorites
    const handleRemoveFromFavorites = (item) => {
        dispatch(removeFavorite(item));
    };

    // Component's JSX rendering
    return (
        <div className="grid grid-cols-1 md:grid-cols-2 lg:grid-cols-3 gap-6 p-4 bg-gray-50 rounded-md">
            {results.map(item => (
                <Item
                    key={item.trackId || item.collectionId}
                    item={item}
                    // Check if current item is in favorites
                    isFavorite={favorites.some(fav => fav.trackId === item.trackid)}
                    addToFavorites={handleAddToFavorites}
                    removeFromFavorites={handleRemoveFromFavorites}
                    onItemClick={() => handleItemClick(item)}
                />
            ))}
        </div>
    );
}

// Exporting the ResultsGrid component
export default ResultsGrid;
