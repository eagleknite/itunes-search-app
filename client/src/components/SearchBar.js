// Required dependencies and components
import { useState, useEffect, useCallback } from 'react';
import { useDispatch, useSelector } from 'react-redux';
import { useNavigate, useSearchParams } from 'react-router-dom';
import { fetchSearchResults, setCurrentSearchTerm } from '../store/searchResultsSlice';
import MediaTypeFilter from './MediaTypeFilter';
import ResultsGrid from './ResultsGrid';
import SortOptions from './SortOptions';

// SearchBar component definition
function SearchBar() {
    // Fetching total results count from the store
    const totalResults = useSelector((state) => state.searchResults.totalResults);
    
    // Component's state for current page and pagination
    const [currentPage, setCurrentPage] = useState(1);
    const totalPages = Math.ceil(totalResults / 50);
    
    // State for search term, media type, and sorting option
    const [term, setTerm] = useState('');
    const [media, setMedia] = useState('all');
    const [sort, setSort] = useState('releaseDate');
    
    // Redux dispatch function
    const dispatch = useDispatch();
    
    // Fetching search results and current search term from the store
    const results = useSelector(state => state.searchResults.items);
    const currentSearchTerm = useSelector(state => state.searchResults.currentSearchTerm);

    // React-router hooks for navigation and search parameters
    const navigate = useNavigate();
    const [searchParams, setSearchParams] = useSearchParams();

    // Effect to update current page from URL parameters
    useEffect(() => {
        const pageParam = searchParams.get('page');
        if (pageParam) setCurrentPage(Number(pageParam));
    }, [searchParams]);

    // Function to handle search
    const handleSearch = useCallback(() => {
        if (term) {
            dispatch(setCurrentSearchTerm(term));  // Setting current search term in store
            dispatch(fetchSearchResults({ term, media, sort, page: currentPage }));  // Dispatching search request
            setSearchParams({ page: currentPage });  // Updating URL with current page
            navigate(`/search?page=${currentPage}`);  // Navigating to updated URL
        }
    }, [dispatch, term, media, sort, currentPage, setSearchParams, navigate]);

    // Effect to trigger search when current page changes
    useEffect(() => {
        if (term) {
            handleSearch();
        }
    }, [currentPage, handleSearch, term]);

    // Effect to trigger search when sort option changes
    useEffect(() => {
        if (term) {
            handleSearch();
        }
    }, [sort, handleSearch, term]);

    // Function to handle form submission
    const handleSubmit = (e) => {
        e.preventDefault();
        handleSearch();
    }

    // Effect to set term from store's current search term
    useEffect(() => {
        if (currentSearchTerm) {
            setTerm(currentSearchTerm);
        }
    }, [currentSearchTerm]);

    return (
        <div className="p-4">
            <div className="max-w-xl mx-auto bg-gray-100 p-4 rounded-md shadow-md">
                <form onSubmit={handleSubmit}>
                    <input
                        type="text"
                        id='searchTerm'
                        value={term}
                        onChange={(e) => setTerm(e.target.value)}
                        placeholder="🔍 Search..."
                        className="w-full p-2 border rounded-md mr-2 mb-2"
                    />
                    <div className="flex justify-between items-center">
                        <MediaTypeFilter selectedMedia={media} onMediaChange={setMedia} />
                        <button onClick={handleSearch} className="bg-blue-500 text-white p-2 rounded-md ml-2 hover:bg-blue-600 transition-colors">Search</button>
                        <SortOptions selectedSort={sort} onSortChange={setSort} />
                    </div>
                </form>
            </div>
            <ResultsGrid />
            {(term || results.length > 0) && (  // Only show the buttons if there's a search term
                <div className="flex justify-center mt-4">
                    <button disabled={currentPage === 1} onClick={() => setCurrentPage(prev => prev - 1)}>Previous</button>
                    {Array.from({ length: totalPages }).map((_, index) => (
                        <button
                            key={index}
                            className={`mx-1 p-2 rounded ${currentPage === index + 1 ? 'bg-blue-500 text-white' : 'bg-white text-blue-500 border border-blue-500'}`}
                            onClick={() => {
                                setCurrentPage(index + 1);
                                handleSearch();
                            }}
                        >
                            {index + 1}
                        </button>
                    ))}
                    <button disabled={currentPage === totalPages} onClick={() => setCurrentPage(prev => prev + 1)}>Next</button>
                </div>
            )}
        </div>
    )
}

export default SearchBar;
