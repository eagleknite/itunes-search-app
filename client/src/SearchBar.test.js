import React from 'react';
import { render, fireEvent, waitFor, screen } from '@testing-library/react';
import { Provider } from 'react-redux';
import configureMockStore from 'redux-mock-store';
import SearchBar from './components/SearchBar';
import SortOptions from './components/SortOptions';
import ResultsGrid from './components/ResultsGrid';
import MediaTypeFilter from './components/MediaTypeFilter';
import FavoritesList from './components/FavoritesList';
import Item from './utils/Item';
import { MemoryRouter } from 'react-router-dom';

const mockStore = configureMockStore();
const store = mockStore({
    searchResults: {
        items: [{
            trackId: 1,
            artworkUrl100: "test.jpg",
            trackName: "Test Track",
            shortDescription: "Description"
        }],
        totalResults: 1
    },
    favorites: [{
        trackId: 1,
        artworkUrl100: "test.jpg",
        trackName: "Test Track",
        shortDescription: "Description"
    }],
    selectedItem: null,
    ui: {
        loading: false,
        error: null
    }
})

test('SearchBar snapshot', () => {
    const { asFragment } = render(
        <Provider store={store}>
            <MemoryRouter>
                <SearchBar />
            </MemoryRouter>
        </Provider>
    );
    expect(asFragment()).toMatchSnapshot();
});

test('SortOptions renders correctly and triggers callback', () => {
    const onSortChange = jest.fn();
    const { getByLabelText } = render(<SortOptions selectedSort="releaseDate" onSortChange={onSortChange} />);
    const select = getByLabelText('Sort By:');
    fireEvent.change(select, { target: { value: 'artistName' } });
    expect(onSortChange).toHaveBeenCalledWith('artistName');
});

test('ResultsGrid renders items', () => {
    const mockData = [
        {
            trackId: 1,
            artworkUrl100: "test.jpg",
            trackName: "Test Track",
            shortDescription: "Description"
        }
    ];
    const { getByText } = render(
        <Provider store={store}>
            <ResultsGrid items={mockData} />
        </Provider>
    );
    expect(getByText('Test Track')).toBeInTheDocument();
});

test('MediaTypeFilter renders media types and triggers callback', () => {
    const onMediaChange = jest.fn();
    const { getByRole } = render(<MediaTypeFilter selectedMedia="all" onMediaChange={onMediaChange} />);
    const select = getByRole('combobox');
    fireEvent.change(select, { target: { value: 'movie' } });
    expect(onMediaChange).toHaveBeenCalledWith('movie');
});

test('FavoritesList renders items and removes favorites', async () => {
    const mockData = [
        {
            trackId: 1,
            artworkUrl100: "test.jpg",
            trackName: "Test Track",
            shortDescription: "Description"
        }
    ];
    const { getByText, queryByText } = render(
        <Provider store={store}>
            <FavoritesList items={mockData} />
        </Provider>
    );
    
    expect(getByText('Test Track')).toBeInTheDocument();
    
    fireEvent.click(getByText('Remove'));
    
    console.log("Favorites in Store after Remove:", store.getState().favorites);  // Logging the updated favorites in store
    
    await waitFor(() => {
        expect(queryByText('Test Track')).toBeNull();
    }, { timeout: 10000 });  // Increasing the timeout
});

test('Item renders correctly and triggers add and remove', () => {
    const mockData = {
        trackId: 1,
        artworkUrl100: "test.jpg",
        trackName: "Test Track",
        shortDescription: "Description"
    };
    const addFavorite = jest.fn();
    const removeFavorite = jest.fn();
    
    // Testing the "Add" functionality
    const { getByText, rerender } = render(
        <Item item={mockData} isFavorite={false} addToFavorites={addFavorite} removeFromFavorites={removeFavorite} />
    );
    fireEvent.click(getByText('Add'));
    expect(addFavorite).toHaveBeenCalledTimes(1);

    // Testing the "Remove" functionality
    rerender(<Item item={mockData} isFavorite={true} addToFavorites={addFavorite} removeFromFavorites={removeFavorite} />);
    fireEvent.click(getByText('Remove'));
    expect(removeFavorite).toHaveBeenCalledTimes(1);
});



