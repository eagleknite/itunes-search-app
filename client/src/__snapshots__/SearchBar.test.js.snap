// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`SearchBar snapshot 1`] = `
<DocumentFragment>
  <div
    class="p-4"
  >
    <div
      class="max-w-xl mx-auto bg-gray-100 p-4 rounded-md shadow-md"
    >
      <form>
        <input
          class="w-full p-2 border rounded-md mr-2 mb-2"
          id="searchTerm"
          placeholder="🔍 Search..."
          type="text"
          value=""
        />
        <div
          class="flex justify-between items-center"
        >
          <select
            class="bg-gray-200 p-2 rounded-md"
          >
            <option
              value="all"
            >
              All
            </option>
            <option
              value="movie"
            >
              Movie
            </option>
            <option
              value="podcast"
            >
              Podcast
            </option>
            <option
              value="music"
            >
              Music
            </option>
            <option
              value="audiobook"
            >
              Audiobook
            </option>
            <option
              value="shortFilm"
            >
              Short Film
            </option>
            <option
              value="tvShow"
            >
              TV Show
            </option>
            <option
              value="software"
            >
              Software
            </option>
            <option
              value="ebook"
            >
              Ebook
            </option>
          </select>
          <button
            class="bg-blue-500 text-white p-2 rounded-md ml-2 hover:bg-blue-600 transition-colors"
          >
            Search
          </button>
          <div
            class="ml-4"
          >
            <label
              class="mr-2"
              for="sortOptions"
            >
              Sort By:
            </label>
            <select
              class="border p-2 rounded-md"
              id="sortOptions"
            >
              <option
                value="releaseDate"
              >
                Release Date
              </option>
              <option
                value="trackPrice"
              >
                Price
              </option>
              <option
                value="primaryGenreName"
              >
                Genre
              </option>
              <option
                value="artistName"
              >
                Artist
              </option>
              <option
                value="collectionName"
              >
                Album
              </option>
            </select>
          </div>
        </div>
      </form>
    </div>
    <div
      class="grid grid-cols-1 md:grid-cols-2 lg:grid-cols-3 gap-6 p-4 bg-gray-50 rounded-md"
    >
      <div
        class="border rounded-md p-3 relative hover:shadow-lg transition-shadow duration-300 cursor-pointer"
      >
        <img
          alt="Test Track"
          class="mx-auto mb-2 w-full"
          src="test.jpg"
        />
        <h3
          class="text-center text-lg font-bold mb-2"
        >
          Test Track
        </h3>
        <p
          class="text-center text-sm mb-2"
        >
          Description
        </p>
        <button
          class="absolute top-2 right-2 bg-blue-500 text-white rounded-full p-1 hover:bg-blue-600"
        >
          Add
        </button>
      </div>
    </div>
    <div
      class="flex justify-center mt-4"
    >
      <button
        disabled=""
      >
        Previous
      </button>
      <button
        class="mx-1 p-2 rounded bg-blue-500 text-white"
      >
        1
      </button>
      <button
        disabled=""
      >
        Next
      </button>
    </div>
  </div>
</DocumentFragment>
`;
